Karl/test.cc

```diff
@@ -312,6 +318,144 @@ TEST(Karl, Deregister) {
+    ASSERT_NO_THROW(
+        status = ParseJSON<unittest_karl_status_t>(
+            HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+    EXPECT_EQ(0u, status.size()) << JSON(status);
+  }
+}
+
+TEST(Karl, DeregisterWithNginx) {
+  current::time::ResetToZero();
+
+  // Run the test only if `karl_nginx_config_file` flag is set.
+  if (FLAGS_karl_nginx_config_file.empty()) {
+    return;
+  }
+
+  const auto stream_file_remover = current::FileSystem::ScopedRmFile(FLAGS_karl_test_stream_persistence_file);
+  const auto storage_file_remover = current::FileSystem::ScopedRmFile(FLAGS_karl_test_storage_persistence_file);
+
+  const current::karl::KarlNginxParameters nginx_parameters(FLAGS_karl_nginx_port,
+                                                            FLAGS_karl_nginx_config_file);
+  const std::string karl_nginx_base_url = "http://localhost:" + current::ToString(FLAGS_karl_nginx_port);
+  const unittest_karl_t karl(FLAGS_karl_test_port,
+                             FLAGS_karl_test_stream_persistence_file,
+                             FLAGS_karl_test_storage_persistence_file,
+                             "/",
+                             karl_nginx_base_url,
+                             nginx_parameters);
+  const current::karl::Locator karl_locator(Printf("http://localhost:%d", FLAGS_karl_test_port));
+
+  // No services registered.
+  {
+    unittest_karl_status_t status;
+    ASSERT_NO_THROW(
+        status = ParseJSON<unittest_karl_status_t>(
+            HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+    EXPECT_EQ(0u, status.size());
+  }
+
+  std::string generator_proxied_status_url;
+  std::string is_prime_proxied_status_url;
+  {
+    const karl_unittest::ServiceGenerator generator(
+        FLAGS_karl_generator_test_port, std::chrono::microseconds(1000), karl_locator);
+    // The `generator` service is registered.
+    {
+      unittest_karl_status_t status;
+      ASSERT_NO_THROW(
+          status = ParseJSON<unittest_karl_status_t>(
+              HTTP(GET(Printf("http://localhost:%d?from=0&full&active_only", FLAGS_karl_test_port))).body));
+      EXPECT_EQ(1u, status.size());
+      ASSERT_TRUE(status.count("127.0.0.1")) << JSON(status);
+      auto& per_ip_services = status["127.0.0.1"].services;
+      EXPECT_EQ(1u, per_ip_services.size());
+      EXPECT_EQ("generator", per_ip_services[generator.ClaireCodename()].service);
+      ASSERT_TRUE(Exists(per_ip_services[generator.ClaireCodename()].url_status_page_proxied));
+      generator_proxied_status_url = Value(per_ip_services[generator.ClaireCodename()].url_status_page_proxied);
+      EXPECT_EQ(karl_nginx_base_url + "/proxied/" + generator.ClaireCodename(), generator_proxied_status_url);
+    }
+
+    // Check that `generator`'s status page is accessible via Nginx.
+    {
+      current::karl::ClaireStatus status;
+      while (true) {
```

Comment:
Need a comment here, along the lines of `must wait for nginx config reload to take effect.`.